File: java-api/src/main/java/com/webmvc/javaapi/controller/ItemController.java
package com.webmvc.javaapi.controller;

import com.webmvc.javaapi.dto.request.ItemSaveRequest;
import com.webmvc.javaapi.dto.response.ApiResponse;
import com.webmvc.javaapi.dto.response.ItemResponse;
import com.webmvc.javaapi.service.ItemService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("item")
@RequiredArgsConstructor
public class ItemController {

    private final ItemService itemService;

    @PostMapping
    public ApiResponse<ItemResponse> saveItem(@RequestBody @Valid ItemSaveRequest itemSaveRequest) {
        ItemResponse responseDto = itemService.saveItem(itemSaveRequest);

        return new ApiResponse<>(responseDto);
    }

    @GetMapping
    public ApiResponse<Page<ItemResponse>> findItems(Pageable pageable, @RequestParam(required = false) String name) {
        Page<ItemResponse> page = itemService.findItems(pageable, name);

        return new ApiResponse<>(page);
    }

    @GetMapping("{id}")
    public ApiResponse<ItemResponse> findItem(@PathVariable Long id) {
        ItemResponse responseDto = itemService.findItem(id);

        return new ApiResponse<>(responseDto);
    }

    @PutMapping("{id}")
    public ApiResponse<ItemResponse> updateItem(
            @PathVariable Long id,
            @RequestBody @Valid ItemSaveRequest itemSaveRequest
    ) {
        ItemResponse responseDto = itemService.updateItem(id, itemSaveRequest);

        return new ApiResponse<>(responseDto);
    }

    @DeleteMapping("{id}")
    public ApiResponse<ItemResponse> deleteItem(@PathVariable Long id) {
        itemService.deleteItem(id);

        return new ApiResponse<>();
    }
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/controller/ItemController.kt
package spring.web.kotlin.controller

import org.springframework.data.domain.Pageable
import org.springframework.web.bind.annotation.*
import spring.web.kotlin.dto.request.ItemSaveRequest
import spring.web.kotlin.dto.response.ApiResponse
import spring.web.kotlin.service.ItemService

@RestController
@RequestMapping("item")
class ItemController(
    private val itemService: ItemService
) {
    @PostMapping
    fun saveItem(@RequestBody itemSaveRequest: ItemSaveRequest) =
        ApiResponse(itemService.saveItem(itemSaveRequest))

    @GetMapping
    fun findItems(pageable: Pageable, @RequestParam(required = false) name: String?) =
        ApiResponse(itemService.findItems(pageable, name))

    @GetMapping("{id}")
    fun findItem(@PathVariable id: Long) = ApiResponse(itemService.findItem(id))

    @PutMapping("{id}")
    fun updateItem(@PathVariable id: Long, @RequestBody itemSaveRequest: ItemSaveRequest) =
        ApiResponse(itemService.updateItem(id, itemSaveRequest))

    @DeleteMapping("{id}")
    fun deleteItem(@PathVariable id: Long): ApiResponse<Unit> {
        itemService.deleteItem(id)

        return ApiResponse()
    }
}